src/ArduinoCloudProperty.cpp

```diff
@@ -0,0 +1,237 @@
+#warning "No RTC available on this architecture - ArduinoIoTCloud will not keep track of local change timestamps ."
+  return 0;
+  #endif
+}
+
+/******************************************************************************
+   CTOR/DTOR
+ ******************************************************************************/
+ArduinoCloudProperty::ArduinoCloudProperty()
+  : _name(""),
+    _permission(Permission::Read),
+    _update_callback_func(nullptr),
+    _sync_callback_func(nullptr),
+    _has_been_updated_once(false),
+    _has_been_modified_in_callback(false),
+    _min_delta_property(0.0f),
+    _min_time_between_updates_millis(0),
+    _last_updated_millis(0),
+    _update_interval_millis(0),
+    _last_local_change_timestamp(0),
+    _last_cloud_change_timestamp(0),
+    _encoder(nullptr),
+    _map_data_list(nullptr) {
+}
+
+/******************************************************************************
+   PUBLIC MEMBER FUNCTIONS
+ ******************************************************************************/
+void ArduinoCloudProperty::init(String const name, Permission const permission) {
+  _name = name;
+  _permission = permission;
+}
+
+ArduinoCloudProperty & ArduinoCloudProperty::onUpdate(UpdateCallbackFunc func) {
+  _update_callback_func = func;
+  return (*this);
+}
+
+ArduinoCloudProperty & ArduinoCloudProperty::onSync(void (*func)(ArduinoCloudProperty & property)) {
+  _sync_callback_func = func;
+  return (*this);
+}
+
+ArduinoCloudProperty & ArduinoCloudProperty::publishOnChange(float const min_delta_property, unsigned long const min_time_between_updates_millis) {
+  _update_policy = UpdatePolicy::OnChange;
+  _min_delta_property = min_delta_property;
+  _min_time_between_updates_millis = min_time_between_updates_millis;
+  return (*this);
+}
+
+ArduinoCloudProperty & ArduinoCloudProperty::publishEvery(unsigned long const seconds) {
+  _update_policy = UpdatePolicy::TimeInterval;
+  _update_interval_millis = (seconds * 1000);
+  return (*this);
+}
+
+bool ArduinoCloudProperty::shouldBeUpdated() {
+  if (!_has_been_updated_once) {
+    return true;
+  }
+
+  if (_has_been_modified_in_callback) {
+    _has_been_modified_in_callback = false;
+    return true;
+  }
+
+  if (_update_policy == UpdatePolicy::OnChange) {
+    return (isDifferentFromCloud() && ((millis() - _last_updated_millis) >= (_min_time_between_updates_millis)));
+  } else if (_update_policy == UpdatePolicy::TimeInterval) {
+    return ((millis() - _last_updated_millis) >= _update_interval_millis);
+  } else {
+    return false;
+  }
+}
+
+void ArduinoCloudProperty::execCallbackOnChange() {
+  if (_update_callback_func != NULL) {
+    _update_callback_func();
+  }
+  if (!isDifferentFromCloud()) {
+    _has_been_modified_in_callback = true;
+  }
+}
+
+void ArduinoCloudProperty::execCallbackOnSync() {
+  if (_sync_callback_func != NULL) {
+    _sync_callback_func(*this);
+  }
+}
+
+void ArduinoCloudProperty::append(CborEncoder *encoder) {
+  _encoder = encoder;
```

Comment:
I believe that storing a pointer to `encoder` as a member variable is both dangerous and unnecessary. Dangerous, because after `cbor_encoder_close_container` the pointer is no longer valid, unnecessary because you can pass it over to its consumers as a parameter. Adding new member variables should always be carefully considered, since they enlarge the state space of the class and therefore the possible error cases to consider.

Reply:
The reason to store the pointer inside the class is to not expose to the derived property classes the visibility of the encoding protocol. In this way the appendAttributesToCloud() of the various CloudBool, CloudInt, ..., CloudLocation and especially the future user defined ones do not need to understand the CBOR stuff or other encoding mechanism.
I removed it by using some macro magic.